buffer/src/main/java/io/netty/buffer/PoolChunk.java

```diff
@@ -222,256 +289,278 @@ private int usage(int freeBytes) {
+        long handle;
+        if (sizeIdx <= arena.smallMaxSizeIdx) {
+            // small
+            handle = allocateSubpage(sizeIdx);
+            if (handle < 0) {
+                return false;
+            }
+            assert isSubpage(handle);
         } else {
-            handle = allocateSubpage(normCapacity);
+            // normal
+            // runSize must be multiple of pageSize
+            int runSize = arena.sizeIdx2size(sizeIdx);
+            handle = allocateRun(runSize);
+            if (handle < 0) {
+                return false;
+            }
         }
 
-        if (handle < 0) {
-            return false;
-        }
-        ByteBuffer nioBuffer = cachedNioBuffers != null ? cachedNioBuffers.pollLast() : null;
-        initBuf(buf, nioBuffer, handle, reqCapacity, threadCache);
+        ByteBuffer nioBuffer = cachedNioBuffers != null? cachedNioBuffers.pollLast() : null;
+        initBuf(buf, nioBuffer, handle, reqCapacity, cache);
         return true;
     }
 
-    /**
-     * Update method used by allocate
-     * This is triggered only when a successor is allocated and all its predecessors
-     * need to update their state
-     * The minimal depth at which subtree rooted at id has some free space
-     *
-     * @param id id
-     */
-    private void updateParentsAlloc(int id) {
-        while (id > 1) {
-            int parentId = id >>> 1;
-            byte val1 = value(id);
-            byte val2 = value(id ^ 1);
-            byte val = val1 < val2 ? val1 : val2;
-            setValue(parentId, val);
-            id = parentId;
-        }
-    }
+    private long allocateRun(int runSize) {
+        int pages = runSize >> pageShifts;
+        int pageIdx = arena.pages2pageIdx(pages);
 
-    /**
-     * Update method used by free
-     * This needs to handle the special case when both children are completely free
-     * in which case parent be directly allocated on request of size = child-size * 2
-     *
-     * @param id id
-     */
-    private void updateParentsFree(int id) {
-        int logChild = depth(id) + 1;
-        while (id > 1) {
-            int parentId = id >>> 1;
-            byte val1 = value(id);
-            byte val2 = value(id ^ 1);
-            logChild -= 1; // in first iteration equals log, subsequently reduce 1 from logChild as we traverse up
-
-            if (val1 == logChild && val2 == logChild) {
-                setValue(parentId, (byte) (logChild - 1));
-            } else {
-                byte val = val1 < val2 ? val1 : val2;
-                setValue(parentId, val);
+        synchronized (runsAvail) {
+            //find first queue which has at least one big enough run
+            int queueIdx = runFirstBestFit(pageIdx);
+            if (queueIdx == -1) {
+                return -1;
+            }
+
+            //get run with min offset in this queue
+            PriorityQueue<Long> queue = runsAvail[queueIdx];
+            long handle = queue.poll();
```

Comment:
can `queue.poll()` ever return `null` ? If so this may produce an `NPE` while unboxing

Reply:
It will not return null, because runFirstBestFit(pageIdx) finds a not empty queue, and synchronized (runsAvail) ensures the queue can not be changed by other threads.